### PR TITLE
Webpack + ES Rewrite

### PR DESCRIPTION
**Pull request by dclstn**
This pull request contains the entire BetterTTV project rewritten in ES modules and Webpack.

**Changes**
- Added [webpack.config.js](https://github.com/dclstn/betterttv/blob/master/webpack.config.js) as new bundling config.
- Added a bunch of webpack packages.
- Added [custom webpack-glob-loader](https://github.com/dclstn/betterttv/blob/master/dev/webpack-import-glob.cjs) for importing betterttv modules.
- Added prettier and updated eslint config to support ES modules and dynamic import statements.
- Editted all files (excluding the custom loader) to support import and export statements.
- Removal of such files like [gulp config](https://github.com/night/betterttv/blob/master/gulpfile.js) and gulp packages.
- Removal of the custom [server.js](https://github.com/night/betterttv/blob/master/dev/server.js) script in replace of [webpack dev server](https://webpack.js.org/configuration/dev-server/).
- Replaced [filesaver.js](https://github.com/night/betterttv/blob/master/src/utils/filesaver.js) with a more up to date module [here](https://www.npmjs.com/package/save-as-file).
- Replaced deprecated [request.js](https://www.npmjs.com/package/request) with [got](https://www.npmjs.com/package/got) + fixed some deprecated URL code.

**Added scripts to package.json**

`npm run start` - start the dev server.
`npm run build` - bundle project into a readable development state.
`npm run build:prod` - bundle project and minify it.
`npm run lint` - to run a linting check on project files.
`npm run pretty` - to run a pretty check on project files.

**Things to note**
- There still may be unused packages in package.json
- Still to rewrite custom loader to ES
- When bundling only the needed files (betterttv.js, betterttv.css and /assets) will be added to /build, previously the entire /src was copied over.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,55 +4,58 @@
   "version": "7.2.92",
   "description": "BetterTTV enhances Twitch with new features, emotes, and more.",
   "main": "betterttv.js",
+  "type": "module",
   "scripts": {
-    "start": "node -e 'require(\"./dev/server.js\")()'"
+    "start": "webpack serve --mode development",
+    "build": "webpack --mode development",
+    "build:prod": "webpack --mode production",
+    "lint": "eslint '**/*.js'",
```
Comment (night) [inline package.json:12]: could we make lint check both eslint and prettier

```
eslint '**/*.js' && prettier --check '**/*.js'
```

File: src/index.js [hunk 2]
```diff
@@ -1,23 +1,14 @@
...
-            'www.twitch.tv',
-            'canary.twitch.tv',
-            'release.twitch.tv',
-            'clips.twitch.tv',
-            'dashboard.twitch.tv',
-            'embed.twitch.tv'
-        ].includes(window.location.hostname) &&
-        !window.location.hostname.endsWith('.release.twitch.tv')
-    ) return;
+    if (!['www.twitch.tv', 'canary.twitch.tv', 'clips.twitch.tv', 'dashboard.twitch.tv', 'embed.twitch.tv'].includes(window.location.hostname)) return;
```
Comment (night) [inline src/index.js:5]: this looks like a breaking change

File: src/index.js [hunk 3]
```diff
@@ -29,27 +20,19 @@
...
-            return `
-                try {
-                    require('${module}');
-                } catch (e) {
-                    debug.error('Failed to ${module}', e.stack);
-                }
-            `;
-        }).join(' ');
-    }});
+    import('./modules/**/index.js');
```
Comment (night) [inline src/index.js:23]: nit: let's make this + the regex check for `await import` instead so it's closer to the actual spec

File: src/index.js [hunk 4]
```diff
@@ -1,23 +1,14 @@
-(() => {
+
+async function main() {
```
Comment (night) [inline src/index.js:2]: do we need this wrapper anymore? iirc with esm we don't anymore

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 5]
```diff
@@ -1,12 +1,14 @@
...
+import watcher from '../../watcher.js';
+import settings from '../../settings.js';
+import storage from '../../storage.js';
+import html from '../../utils/html.js';
+import twitch from '../../utils/twitch.js';
+import cdn from '../../utils/cdn.js';
+import moment from 'moment';
+import regex from '../../utils/regex.js';
+
+const escapeRegExp = regex.escape;
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:11]: instead of this, let's change the exports in the regex util to expose the function itself (rather than in default)

File: src/watcher.js [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
 const CLIPS_HOSTNAME = 'clips.twitch.tv';
 
 class Watcher extends SafeEventEmitter {
     setup() {
-        require('./watchers/channel')(this);
+        channelWatcher(this);
 
         if (window.location.hostname === CLIPS_HOSTNAME) {
-            require('./watchers/clips')(this);
+            clipsWatcher(this);
```
Comment (night) [inline src/watcher.js:18]: we intentionally lazy load to prevent loading code inapplicable to the page. let's revert this back to dynamic imports

File: webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
+import { CleanWebpackPlugin } from 'clean-webpack-plugin';
```
Comment (night) [inline webpack.config.js:1]: it doesn't look like prettier has been ran on many of these files (brackets should not have spaces according to the rules)

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    const result = {};
+    for (const emojiData of Object.values(emojis)) {
+        const char = convertEmojiToolkitCodePointToChar(emojiData.code_points.fully_qualified);
+        const data = {
+            char,
+            slug: emojiData.shortname.replace(/:/g, ''),
+        };
+        result[data.slug] = data;
+        for (const alternativeShortName of emojiData.shortname_alternates) {
+            result[alternativeShortName.replace(/:/g, '')] = data;
```
Comment (night) [inline webpack.config.js:47]: you will need to merge in master or rebase your changes off master to pull in recent changes

File: webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    const { version } = JSON.parse(await fs.readFile('./package.json'));
+    const emotes = JSON.parse(await fs.readFile('./node_modules/emoji-toolkit/emoji.json'));
+
+    return {
+        devServer: {
+            contentBase: path.resolve('./build'),
+            compress: true,
+            port: PORT,
+            after: function(app) {
+                app.get('*', async(req, res) => {
```
Comment (night) [inline webpack.config.js:69]: async looks unnecessary here

File: webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+    const { version } = JSON.parse(await fs.readFile('./package.json'));
+    const emotes = JSON.parse(await fs.readFile('./node_modules/emoji-toolkit/emoji.json'));
+
+    return {
+        devServer: {
+            contentBase: path.resolve('./build'),
+            compress: true,
+            port: PORT,
+            after: function(app) {
```
Comment (night) [inline webpack.config.js:68]: nit: arrow function

**Review by night — CHANGES_REQUESTED** [review]

File: src/modules/chat_highlight_blacklist_keywords/index.js (hunk 5 above)
Comment (dclstn) [reply src/modules/chat_highlight_blacklist_keywords/index.js:11]: utils/regix.js exports 2 functions, is the other-one unused?

File: webpack.config.js (hunk 10 above)
Comment (dclstn) [reply webpack.config.js]: changed to arrow func

File: package.json (hunk 1 above)
Comment (dclstn) [reply package.json]: scripts have been combined

File: webpack.config.js (hunk 9 above)
Comment (dclstn) [reply webpack.config.js]: removed stray async

File: src/index.js (hunk 2 above)
Comment (dclstn) [reply src/index.js]: has been reverted back

File: src/index.js (hunk 3 above)
Comment (dclstn) [reply src/index.js]: changes to await import

File: src/index.js (hunk 4 above)
Comment (dclstn) [reply src/index.js]: removed wrapper

File: src/watcher.js (hunk 6 above)
Comment (dclstn) [reply src/watcher.js]: should be using dynamic imports now

File: webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+import fs from 'fs/promises';
+import path from 'path';
+import {createRequire} from 'module';
+import globPkg from 'glob';
+import TerserPlugin from 'terser-webpack-plugin';
+import postcssUrl from 'postcss-url';
+import got from 'got';
+import CssMinimizerPlugin from 'css-minimizer-webpack-plugin';
+
+const git = createRequire(import.meta.url)('git-rev-sync');
```
Comment (night) [inline webpack.config.js:16]: why is this needed instead of a normal import?

File: webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        output: {
+            filename: '[name].js',
+            path: path.resolve('./build'),
+        },
+        module: {
+            rules: [
+                {
+                    test: /\.js$/,
+                    enforce: 'pre',
+                    loader: path.resolve('./dev/webpack-import-glob.cjs'),
```
Comment (night) [inline webpack.config.js:90]: loader documentation shows they support esm. can we move this cjs file to esm?

File: src/modules/chat_highlight_blacklist_keywords/index.js (hunk 5 above)
Comment (night) [reply src/modules/chat_highlight_blacklist_keywords/index.js:11]: looks like it

Comment (dclstn) [reply src/modules/chat_highlight_blacklist_keywords/index.js]: done, also removed unused code

File: .travis.yml [hunk 13]
```diff
@@ -2,16 +2,15 @@ language: node_js
...
 node_js:
-- '12.10'
+- '15.10'
 before_script:
 - which gulp || npm install -g gulp
-script: gulp dist
+script: npm run lint && npm run build:prod
 notifications:
   email: false
 cache:
```
Comment (night) [inline .travis.yml:11]: we can apparently set this to

```
cache: npm
```

File: webpack.config.js (hunk 11 above)
Comment (dclstn) [reply webpack.config.js]: Some ES modules are not able to be imported in Node 14+, you find the solution to this [here](https://dev.to/jimthedev/typeerror-cannot-read-property-filename-of-undefined-in-es-modules-in-node-14-and-greater-6ia)

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 14]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
+        case 3: // CurrentUserHighlight
+          return node.content;
+        case 4: // Mention
+          return node.content.recipient;
+        case 5: // Link
+          return node.content.url;
+        case 6: // Emote
+          return node.content.alt;
+        default:
+          return false;
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:168]: why would this return false?

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 15]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
-
-        if (fromContainsKeyword(blacklistUsers, from) || messageContainsKeyword(blacklistKeywords, from, message)) {
-            return this.markBlacklisted($message);
-        }
+    if (fromContainsKeyword(highlightUsers, from) || messageContainsKeyword(highlightKeywords, from, message)) {
+      this.markHighlighted($message);
 
-        if (fromContainsKeyword(highlightUsers, from) || messageContainsKeyword(highlightKeywords, from, message)) {
-            this.markHighlighted($message);
+      if (isReply($message)) return false;
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:249]: revert this.

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 16]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
+    changeKeywords(HIGHLIGHT_KEYWORD_PROMPT, 'highlightKeywords');
+  }
+
+  onMessage($message, {user, timestamp, messageParts}) {
+    const from = user.userLogin;
+    const message = messageTextFromAST(messageParts);
+    const date = new Date(timestamp);
+
+    if (fromContainsKeyword(blacklistUsers, from) || messageContainsKeyword(blacklistKeywords, from, message)) {
+      return this.markBlacklisted($message);
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:243]:
```
this.markBlacklisted($message);
return;
```

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 17]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
-    markHighlighted($message) {
-        $message.addClass('bttv-highlighted');
+    if (fromContainsKeyword(highlightUsers, from) || messageContainsKeyword(highlightKeywords, from, messageContent)) {
+      return this.markHighlighted($message);
     }
 
-    markBlacklisted($message) {
-        $message.attr('style', 'display: none !important;');
-    }
+    return false;
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:277]: revert this

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 18]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
+    const messageContent = `${$messageContent.text().replace(/^:/, '')} ${emotes.join(' ')}`;
+
+    if (fromContainsKeyword(blacklistUsers, from) || messageContainsKeyword(blacklistKeywords, from, messageContent)) {
+      return this.markBlacklisted($message);
     }
 
-    markHighlighted($message) {
-        $message.addClass('bttv-highlighted');
+    if (fromContainsKeyword(highlightUsers, from) || messageContainsKeyword(highlightKeywords, from, messageContent)) {
+      return this.markHighlighted($message);
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:274]: revert this.

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 19]
```diff
@@ -42,266 +44,278 @@ const pinnedHighlightTemplate = ({timestamp, from, message}) => `
...
+    const $from = $message.find(VOD_CHAT_FROM_SELECTOR);
+    const from = ($from.attr('href') || '').split('?')[0].split('/').pop();
+    const $messageContent = $message.find(VOD_CHAT_MESSAGE_SELECTOR);
+    const emotes = Array.from($messageContent.find(VOD_CHAT_MESSAGE_EMOTE_SELECTOR)).map((emote) =>
+      emote.getAttribute('alt')
+    );
+    const messageContent = `${$messageContent.text().replace(/^:/, '')} ${emotes.join(' ')}`;
+
+    if (fromContainsKeyword(blacklistUsers, from) || messageContainsKeyword(blacklistKeywords, from, messageContent)) {
+      return this.markBlacklisted($message);
```
Comment (night) [inline src/modules/chat_highlight_blacklist_keywords/index.js:270]:
```
this.markBlacklisted($message);
return;
```

File: src/modules/emotes/personal-emotes.js [hunk 20]
```diff
@@ -1,122 +1,124 @@
...
 
-    getEligibleEmote(code, user) {
-        if (!user) return;
+    return [...emotes.values()];
+  }
 
-        const emotes = this.emotes.get(user.name);
-        if (!emotes) return;
+  getEligibleEmote(code, user) {
+    if (!user) return false;
```
Comment (night) [inline src/modules/emotes/personal-emotes.js:44]: it looks like this should return null

File: src/modules/emotes/personal-emotes.js [hunk 21]
```diff
@@ -1,122 +1,124 @@
...
 
-        return emotes.get(code);
-    }
+    const emotes = this.emotes.get(user.name);
+    if (!emotes) return false;
```
Comment (night) [inline src/modules/emotes/personal-emotes.js:47]: it looks like this should return null

File: src/socket-client.js [hunk 22]
```diff
@@ -18,103 +18,104 @@ let attempts = 1;
...
 
-        socket.onerror = socket.onclose = () => {
-            if (state === CONNECTION_STATES.DISCONNECTED) return;
+      if (joinedChannels.length) {
+        joinedChannels.forEach((c) => this.joinChannel(c));
+      }
+    };
 
-            debug.log('SocketClient: Disconnected from server');
+    // eslint-disable-next-line no-multi-assign
```
Comment (night) [inline src/socket-client.js:48]: lets not disable this..

```
socket.onclose = () => ...
socket.onerror = socket.onclose
```

File: src/utils/cdn.js [hunk 23]
```diff
@@ -1,18 +1,18 @@
-const $ = require('jquery');
-const debug = require('./debug');
+import $ from 'jquery';
+import debug from './debug.js';
 
-const CDN_ENDPOINT = process.env.CDN_ENDPOINT;
+const {CDN_ENDPOINT} = process.env;
```
Comment (night) [inline src/utils/cdn.js:4]: this .. may not work? iirc webpack replaces strings like `process.env.*`. have you tested this change?

File: src/utils/twitch-api.js [hunk 24]
```diff
@@ -1,68 +1,73 @@
...
+import $ from 'jquery';
+import querystring from 'querystring';
 
 const API_ENDPOINT = 'https://api.twitch.tv/v5/';
 const GQL_ENDPOINT = 'https://gql.twitch.tv/gql';
 const CLIENT_ID = '6x8avioex0zt85ht6py4sq55z6avsea';
 
 let accessToken;
 
+class HTTPError extends Error {
```
Comment (night) [inline src/utils/twitch-api.js:10]: rather than redefining this, is there a way to share this code?

File: src/watchers/chat.js [hunk 25]
```diff
@@ -1,73 +1,74 @@
...
-    return twitchHandleMessage.call(this, ...arguments);
+  if (message && typeof message.type === 'number') {
+    let isPrevented = false;
+    watcher.emit('chat.message.handler', {
+      message,
+      preventDefault: () => {
+        isPrevented = true;
+      },
+    });
+    if (isPrevented) return false;
```
Comment (night) [inline src/watchers/chat.js:19]: this should probably return null

File: src/watchers/chat.js [hunk 26]
```diff
@@ -1,73 +1,74 @@
...
+  }
+
+  return twitchHandleMessage.call(this, message);
```
Comment (night) [inline src/watchers/chat.js:22]: this is intentionally using arguments since twitch could break this at any point. for forwards compatibility, we could use spreading on the function declaration, and spreading here

File: src/watchers/clips.js [hunk 27]
```diff
@@ -1,28 +1,28 @@
...
+import $ from 'jquery';
+import twitch from '../utils/twitch.js';
+import domObserver from '../observers/dom.js';
 
-module.exports = watcher => {
-    domObserver.on('.tw-mg-b-1', (node, isConnected) => {
-        if (!isConnected || !node.parentNode.classList.contains('clips-chat-replay')) return;
-        watcher.emit('clips.message', $(node));
-    });
+export default (watcher) => {
```
Comment (night) [inline src/watchers/clips.js:5]: lets use a named function rather than an anonymous arrow function

File: src/watchers/chat.js [hunk 28]
```diff
@@ -1,73 +1,74 @@
...
-    messageHandlerAPI._bttvMessageHandlerPatched = PATCHED_SYMBOL;
-    twitchHandleMessage = handleMessage;
+  messageHandlerAPI.handleMessage = bttvHandleMessage;
+  messageHandlerAPI._bttvMessageHandlerPatched = PATCHED_SYMBOL;
+  twitchHandleMessage = handleMessage;
 }
 
-module.exports = watcher_ => {
-    watcher = watcher_;
+export default (watcher_) => {
```
Comment (night) [inline src/watchers/chat.js:42]: lets use a named function rather than an anonymous arrow function

File: src/watchers/channel.js [hunk 29]
```diff
@@ -1,31 +1,32 @@
...
+      channelEmotes: [],
+      sharedEmotes: [],
+      status: error.status || 0,
+    }))
+    .then((data) => watcher.emit('channel.updated', data));
 }
 
-module.exports = watcher_ => {
-    watcher = watcher_;
+export default (watcher_) => {
```
Comment (night) [inline src/watchers/channel.js:26]: lets use a named function rather than an anonymous arrow function

File: src/watchers/conversations.js [hunk 30]
```diff
@@ -1,20 +1,20 @@
...
-const twitch = require('../utils/twitch');
-const domObserver = require('../observers/dom');
+import $ from 'jquery';
+import twitch from '../utils/twitch.js';
+import domObserver from '../observers/dom.js';
 
-module.exports = watcher => {
-    domObserver.on('.whispers-thread', (node, isConnected) => {
-        if (!isConnected) return;
+export default (watcher) => {
```
Comment (night) [inline src/watchers/conversations.js:5]: lets use a named function rather than an anonymous arrow function

File: src/watchers/routes.js [hunk 31]
```diff
@@ -11,176 +11,179 @@ let currentChatReference;
...
+      waitForLoad('chat').then(() => watcher.emit('load.chat'));
+      break;
+    default:
+      break;
+  }
 }
 
-module.exports = watcher_ => {
-    watcher = watcher_;
+export default (watcher_) => {
```
Comment (night) [inline src/watchers/routes.js:168]: lets use a named function rather than an anonymous arrow function

File: src/modules/chat_deleted_messages/index.js [hunk 32]
```diff
@@ -1,91 +1,93 @@
...
+      return false;
     }
+    const messages = findAllUserMessages(name);
+    messages.forEach((message) => {
+      const $message = $(message);
+      if (hideDeletedMessages) {
+        $message.hide();
+      } else if (showDeletedMessages) {
+        $message.toggleClass(CHAT_LINE_DELETED_CLASS, true);
+        $message.find(CHAT_LINE_LINK_SELECTOR).each(() => {
```
Comment (night) [inline src/modules/chat_deleted_messages/index.js:82]: this change breaks jQuery, where `this` corresponds to the current context

File: src/constants.js [hunk 33]
```diff
@@ -1,67 +1,64 @@
-exports.bots = [
-    'nightbot',
-    'moobot',
-    'xanbot'
-];
+export default {
```
[the remark this answers is not in the document]
Comment (dclstn) [reply src/constants.js]: file has been removed

File: src/modules/chat_highlight_blacklist_keywords/index.js [hunk 34]
```diff
@@ -1,12 +1,13 @@
...
-const {escape: escapeRegExp} = require('../../utils/regex');
+import $ from 'jquery';
+import moment from 'moment';
+import watcher from '../../watcher.js';
+import settings from '../../settings.js';
+import storage from '../../storage.js';
+import html from '../../utils/html.js';
+import twitch from '../../utils/twitch.js';
+import cdn from '../../utils/cdn.js';
+// eslint-disable-next-line import/named
```
[the remark this answers is not in the document]
Comment (dclstn) [reply src/modules/chat_highlight_blacklist_keywords/index.js]: removed stray eslint ignore line

File: package.json [hunk 35]
```diff
@@ -4,55 +4,60 @@
   "version": "7.2.94",
   "description": "BetterTTV enhances Twitch with new features, emotes, and more.",
   "main": "betterttv.js",
+  "type": "module",
   "scripts": {
-    "start": "node -e 'require(\"./dev/server.js\")()'"
+    "start": "webpack serve --mode development",
+    "build": "webpack --mode development",
+    "build:prod": "webpack --mode production",
+    "lint": "eslint '**/*.js' && prettier --check '**/*.js'"
```
[the remark this answers is not in the document]
Comment (dclstn) [reply package.json:12]: added `&& prettier --check '**/*.css'` check

File: src/modules/chat_nicknames/index.js [hunk 36]
```diff
@@ -1,27 +1,27 @@
...
+  constructor() {
+    nicknames = storage.get('nicknames') || {};
+  }
 
-    set(name) {
-        let nickname = prompt(`Enter the updated nickname for ${name} (Leave blank to reset)`, nicknames[name] || name);
-        if (nickname === null) return;
+  set(name) {
+    let nickname = prompt(`Enter the updated nickname for ${name} (Leave blank to reset)`, nicknames[name] || name);
+    if (nickname === null) return false;
```
[the remark this answers is not in the document]
Comment (dclstn) [reply src/modules/chat_nicknames/index.js]: now returns null

File: .eslintrc [hunk 37]
```diff
@@ -1,151 +1,34 @@
...
+    "class-methods-use-this": "off",
+    "no-plusplus": "off",
+    "no-empty": "off",
+    "max-classes-per-file": "off",
+    "import/prefer-default-export": "off",
+    "no-unused-expressions": "off",
+    "no-alert": "off",
+    "no-restricted-properties": "off",
+    "no-case-declarations": "off",
+    "no-prototype-builtins": "off",
```
[the remark this answers is not in the document]
Comment (dclstn) [reply .eslintrc]: removed rule + replaced instances of hasOwnProperty

File: src/modules/chat_highlight_blacklist_keywords/index.js (hunk 14 above)
Comment (dclstn) [reply src/modules/chat_highlight_blacklist_keywords/index.js:167]: .map expects a return, and returning false is essentially doing a continue

Comment (night) [reply src/modules/chat_highlight_blacklist_keywords/index.js:167]: empty string goes here

File: package.json [hunk 38]
```diff
@@ -4,55 +4,62 @@
...
-    "gulp-if": "^3.0.0",
-    "gulp-json-transform": "^0.4.7",
-    "gulp-postcss": "^9.0.0",
-    "gulp-rename": "^2.0.0",
-    "gulp-sourcemaps": "^3.0.0",
-    "gulp-tar": "^3.1.0",
-    "gulp-uglify": "^3.0.2",
+    "glob": "^7.1.6",
+    "got": "^11.8.2",
+    "i": "^0.3.6",
```
Comment (night) [inline package.json:38]: what is this?

File: package.json [hunk 39]
```diff
@@ -4,55 +4,62 @@
...
+    "import-glob-loader": "^1.1.0",
     "jquery": "^3.6.0",
+    "json-loader": "^0.5.7",
     "lodash.debounce": "^4.0.8",
     "lodash.throttle": "^4.1.1",
+    "mini-css-extract-plugin": "^1.4.1",
     "moment": "^2.29.1",
+    "npm": "^7.10.0",
```
Comment (night) [inline package.json:46]: why is this here?

File: src/modules/chat_commands/index.js [hunk 40]
```diff
@@ -1,76 +1,78 @@
...
+  const hours = Math.floor(s / 3600) - days * 24;
+  const minutes = Math.floor(s / 60) - days * 1440 - hours * 60;
+  const seconds = s - days * 86400 - hours * 3600 - minutes * 60;
 
-    return (days > 0 ? days + ' day' + (days === 1 ? '' : 's') + ', ' : '') +
-           (hours > 0 ? hours + ' hour' + (hours === 1 ? '' : 's') + ', ' : '') +
-           (minutes > 0 ? minutes + ' minute' + (minutes === 1 ? '' : 's') + ', ' : '') +
-           seconds + ' second' + (seconds === 1 ? '' : 's');
+  return `${
+    (days > 0 ? `${days} day${days === 1 ? '' : 's'}, ` : '') +
```
[the remark this answers is not in the document]
Comment (night) [reply src/modules/chat_commands/index.js:39]:
```
return `${days > 0 ? `${days} day${days === 1 ? '' : 's'}, ` : ''} ${hours > 0 ? `${hours} hour${hours === 1 ? '' : 's'}, ` : ''}  ${minutes > 0 ? `${minutes} minute${minutes === 1 ? '' : 's'}, ` : ''} ${seconds} seconds${seconds === 1 ? '' : 's'}`;
```

File: src/socket-client.js [hunk 41]
```diff
@@ -18,103 +18,105 @@ let attempts = 1;
...
 
-    joinChannel(name) {
-        if (!joinedChannels.includes(name)) {
-            joinedChannels.push(name);
-        }
-        this.send('join_channel', {name});
-        this.broadcastMe(name);
-    }
+    // eslint-disable-next-line no-restricted-properties
+    setTimeout(() => this.connect(), Math.random() * (Math.pow(2, attempts) - 1) * 30000);
```
Comment (night) [inline src/socket-client.js:87]: pls fix Math.pow

File: src/modules/chat_deleted_messages/index.js [hunk 42]
```diff
@@ -1,91 +1,93 @@
...
+      return false;
     }
+    const messages = findAllUserMessages(name);
+    messages.forEach((message) => {
+      const $message = $(message);
+      if (hideDeletedMessages) {
+        $message.hide();
+      } else if (showDeletedMessages) {
+        $message.toggleClass(CHAT_LINE_DELETED_CLASS, true);
+        $message.find(CHAT_LINE_LINK_SELECTOR).each(function func() {
```
Comment (night) [inline src/modules/chat_deleted_messages/index.js:82]: let's not call this `func` ..

**Review by night — APPROVED** [review]
